# Custom exercice: Using TAP to get usable tables for LePHARE

**Learning objective:** Create 2 tables of LSST objects: one for galaxies, one for stars

**LSST data products:** `Object` table

**Packages:** `lsst.rsp.get_tap_service`

## 1. Introduction

TAP (Table Access Protocol) provides standardized access to catalog data for discovery, search, and retrieval.
Full [documentation for TAP](http://www.ivoa.net/documents/TAP) is provided by the
International Virtual Observatory Alliance (IVOA).

### 1.1. TAP glossary

* **schema** - Database terminology for the abstract design that represents the storage of data in a database.
* **TAP schema** - The set of tables that describe the data tables and their columns.
* **table collection** - A collection of tables (catalogs), e.g., for a given data release.
* **table** - A collection of related data held in a table format in a database. 
* **query** - A string formatted in ADQL that selects data from a table, with contraints if desired. 
* **results** - The output of the TAP service's search method when a query is passed.


### 1.2. Reminder: ADQL statements

The [documentation for ADQL](http://www.ivoa.net/documents/latest/ADQL.html)
includes more information about syntax, keywords, operators, functions, and so on.
ADQL is similar to SQL (Structured Query Langage).

A typical ADQL statement has at least three components:

```
SELECT <columns> FROM <catalog> WHERE <constraints>
```

where
 * `<columns>` is a comma-separated list of the columns to return,
 * `<catalog>` is the name of the catalog to retreive data from, and
 * `<constraints>` imposes a restriction that only rows with column values that meet the constraints are returned.

For example, say there is a catalog called "mysurveydata" with 5 columns, "col1", "col2", and so on.
The ADQL statement:

```
SELECT col3, col4, col5 FROM mysurveydata WHERE col1 > 0.5 AND col5 < 10
```

would 
return a table that has three columns, and as many rows as meet both of the restrictions in the `WHERE` statement.

### 1.3. Import packages


In [5]:
import numpy as np
import matplotlib.pyplot as plt

import lsst.geom as geom

from lsst.rsp import get_tap_service
from lsst.rsp.utils import get_pyvo_auth
from lsst.rsp.service import get_siav2_service


## 2. Initialisation of the query

We will be looking at the DP1 catalog schema table to pick columns that would be helpful to identify the source-type. Then we will make our final query to get the right data.

### 2.1 First look at the schema tables.

In [6]:
service = get_tap_service("tap")
assert service is not None

In [7]:
results = service.search('SELECT * FROM tap_schema.schemas')

In [8]:
results.to_table()

<Table length=4>
   schema_name    utype  ... schema_index
      str64       str512 ...    int32    
----------------- ------ ... ------------
dp02_dc2_catalogs        ...            2
              dp1        ...            0
             ivoa        ...            1
       tap_schema        ...       100000

In [9]:
del results

We will obviously chose dp1, but we need to choose from which catalog we will retrieve data. Thus we look at the tap_schema.tables category.

In [19]:
query = "SELECT * FROM tap_schema.tables " \
        "WHERE tap_schema.tables.schema_name = 'dp1'" \
        "ORDER BY table_index ASC"
results = service.search(query).to_table()
results

<Table length=12>
schema_name          table_name         ... table_index
   str512              str64            ...    int32   
----------- --------------------------- ... -----------
        dp1                  dp1.Object ...          10
        dp1                  dp1.Source ...          20
        dp1            dp1.ForcedSource ...          30
        dp1               dp1.DiaObject ...          40
        dp1               dp1.DiaSource ...          50
        dp1 dp1.ForcedSourceOnDiaObject ...          60
        dp1            dp1.CoaddPatches ...          70
        dp1                   dp1.Visit ...          80
        dp1                dp1.CcdVisit ...          90
        dp1                  dp1.MPCORB ...         100
        dp1                dp1.SSObject ...         110
        dp1                dp1.SSSource ...         120

In [20]:
del query, results

Lets is use the object catalog for this time.

### 2.2 Pick right columns

In the dp1.Object catalog, we need to identify which columns should be useful for LePHARE and later galax

In [21]:
query = "SELECT column_name, datatype, description, unit " \
        "FROM tap_schema.columns " \
        "WHERE table_name = 'dp1.Object'"
results = service.search(query).to_table()
results

<Table length=1296>
             column_name              datatype ...  unit 
                str64                  str64   ... str64 
------------------------------------- -------- ... ------
                            coord_dec   double ...    deg
                         coord_decErr    float ...    deg
                             coord_ra   double ...    deg
                     coord_ra_dec_Cov    float ... deg**2
                          coord_raErr    float ...    deg
                       deblend_failed  boolean ...       
               deblend_incompleteData  boolean ...       
               deblend_isolatedParent  boolean ...       
                   deblend_iterations      int ...       
                                  ...      ... ...    ...
        z_psfModel_TwoGaussian_n_iter      int ...       
z_psfModel_TwoGaussian_no_inputs_flag  boolean ...       
  z_psfModel_TwoGaussian_unknown_flag  boolean ...       
                                 z_ra   double ...    deg
                         z_ra_dec_Cov    float ... deg**2
                              z_raErr    float ...    deg
                         z_sersicFlux    float ...    nJy
                      z_sersicFluxErr    float ...    nJy
                   z_sizeExtendedness    float ...       
              z_sizeExtendedness_flag  boolean ...

In [22]:
search_string = 'ModelMag' #g_cModelMag, g_cModelMagErr
# search_string = 'g_psfFlux'
for cname in results['column_name']:
    if cname.find(search_string) > -1:
        print(cname)

g_cModelMag
g_cModelMagErr
i_cModelMag
i_cModelMagErr
r_cModelMag
r_cModelMagErr
u_cModelMag
u_cModelMagErr
y_cModelMag
y_cModelMagErr
z_cModelMag
z_cModelMagErr


In [23]:
del results

len(results) = 1296 --> we won't need every row

### 2.3. Reminder: useful keywords

We will use a few useful category of keywords to retrieve particular data type:
 * `coord`: ra, dec, errors. Coordinates of the object.
 * `(de)blend`: blendness category includes flag characterizing the mixing of several sources. Perhaps it should be taken into account if blendness and extendedness appeared both as True for a single object
 * `magnitude`: magnitude categories include values associated to ugrizy mag, like numerical value, errors and extendedness. 
 * `{band}_extendedness` where band = ugrizy. This is probably the most useful keywords to identify the source or object type (galaxy or star)

## 3. Trial in the ecdfs region

In [49]:
target_ra = 53.195
target_dec = -27.703
target_r = 1

We won't take into account the AGNs (present in the DIAobject catalog). We will only look at galaxies and stars in the object catalog. To differentiate them efficiently within the LSST catalog, we establish a selection criterion : if 3 bands or more show {band}_extendedness = 1, we'll consider it as a galaxies. Unless the sources will be flagged as stars. But that something that will be useful later.

### 3.1 Full useful parameters table

The first table include every parameters that could be useful in star-galaxies separation, not differentiating the source types.

In [35]:
query = "SELECT objectId, coord_ra, coord_dec, " + \
        "u_cModelMag, u_cModelMagErr, u_extendedness, " + \
        "g_cModelMag, g_cModelMagErr, g_extendedness, " + \
        "r_cModelMag, r_cModelMagErr, r_extendedness, " + \
        "i_cModelMag, i_cModelMagErr, i_extendedness, " + \
        "z_cModelMag, z_cModelMagErr, z_extendedness, " + \
        "y_cModelMag, y_cModelMagErr, y_extendedness, refBand " + \
        "FROM dp1.Object AS obj " + \
        "WHERE (i_cModelFlux/i_cModelFluxErr > 15) AND " + \
        "(u_cModelFlux/u_cModelFluxErr > 5) AND " + \
        "(shape_flag = 0) AND " + \
        "CONTAINS(POINT('ICRS', coord_ra, coord_dec), " + \
        "CIRCLE('ICRS',"+str(target_ra)+","+str(target_dec)+", "+str(target_r)+")) = 1 "


In [36]:
job = service.submit_job(query)
job.run()
job.wait(phases=['COMPLETED', 'ERROR'])
print('Job phase is', job.phase)

Job phase is COMPLETED


In [37]:
results = job.fetch_result()
tab = results.to_table()
tab

<Table length=19641>
     objectId           coord_ra      ... y_extendedness refBand
                          deg         ...                       
      int64             float64       ...    float32       str1 
------------------ ------------------ ... -------------- -------
611257134226610676  53.10476748835234 ...             --       i
611257134226610953 53.121419521326864 ...             --       i
611257134226611953 53.128689161074604 ...             --       g
611257134226611840 53.092374606157875 ...             --       i
611257134226611818  53.11147971348439 ...             --       i
611257134226611819 53.110912371550945 ...             --       g
611257134226611188  53.10320303014881 ...             --       i
611257134226610732  53.09457245686273 ...             --       i
611257134226604870 53.130886640543594 ...             --       i
               ...                ... ...            ...     ...
611256447031851972  53.09556300600793 ...            1.0       i
611256447031854328  53.18492555450073 ...             --       i
611256447031852290 53.184022249802325 ...             --       i
611256447031854244  53.13652321185081 ...             --       i
611256447031854336  53.23027919696859 ...             --       i
611256447031852273  53.19591265610111 ...            1.0       i
611256447031852278   53.2610058307189 ...             --       i
611256447031854243  53.14358991948668 ...            1.0       i
611256447031852262  53.13532738250357 ...            1.0       y

### 3.2 Galaxy table

This is already in the short MEME lephare format.

In [71]:
query = "SELECT objectId, " + \
        "u_cModelMag, u_cModelMagErr, " + \
        "g_cModelMag, g_cModelMagErr, " + \
        "r_cModelMag, r_cModelMagErr, " + \
        "i_cModelMag, i_cModelMagErr, " + \
        "z_cModelMag, z_cModelMagErr, " + \
        "y_cModelMag, y_cModelMagErr " + \
        "FROM dp1.Object AS obj " + \
        "WHERE (i_cModelFlux/i_cModelFluxErr > 15) AND " + \
        "(u_cModelFlux/u_cModelFluxErr > 5) AND " + \
        "(shape_flag = 0) AND " + \
        "(g_extendedness = 1) AND " + \
        "(r_extendedness = 1) AND " + \
        "(i_extendedness = 1) AND " + \
        "(z_extendedness = 1) AND " + \
        "CONTAINS(POINT('ICRS', coord_ra, coord_dec), " + \
        "CIRCLE('ICRS',"+str(target_ra)+","+str(target_dec)+", "+str(target_r)+")) = 1 "

In [72]:
job = service.submit_job(query)
job.run()
job.wait(phases=['COMPLETED', 'ERROR'])
print('Job phase is', job.phase)

Job phase is COMPLETED


In [73]:
results = job.fetch_result()
tab = results.to_table()
tab

<Table length=27605>
     objectId      u_cModelMag u_cModelMagErr ... y_cModelMag y_cModelMagErr
                       mag          mag       ...     mag          mag      
      int64          float32      float32     ...   float32      float32    
------------------ ----------- -------------- ... ----------- --------------
611257134226611432     23.0569        0.19202 ...          --             --
611257134226611457     22.7062       0.139444 ...          --             --
611257134226610818     22.9006       0.187118 ...          --             --
611257134226611169      21.231      0.0615021 ...          --             --
611257134226611170     21.6475      0.0898182 ...          --             --
611256447031859676     23.3172       0.197396 ...          --             --
611256447031859643     22.5311       0.147581 ...          --             --
611256447031840750     22.5936       0.134958 ...          --             --
611256447031859493     22.4865       0.142913 ...          --             --
               ...         ...            ... ...         ...            ...
611254454167027880     25.0423       0.193343 ...     24.1036       0.367641
611254454167027004     23.1624      0.0547045 ...     19.9861      0.0166826
611254454167027263     24.0687       0.104428 ...     22.7793       0.136745
611254454167027857     23.8865      0.0995115 ...     21.4292      0.0668891
611254454167027855     23.5332       0.104836 ...     22.1017       0.101175
611254454167026730     24.7268       0.153275 ...     21.1484      0.0280449
611254454167027872     24.6866       0.160193 ...     23.6027       0.262752
611254454167027849      24.148      0.0907456 ...     22.3678      0.0819796
611254454167027838     23.5888      0.0748752 ...     21.0938      0.0304936

In [74]:
job.delete()
del query, results, tab

### 3.2 Star table

In [63]:
query = "SELECT objectId, " + \
        "u_cModelMag, u_cModelMagErr, " + \
        "g_cModelMag, g_cModelMagErr, " + \
        "r_cModelMag, r_cModelMagErr, " + \
        "i_cModelMag, i_cModelMagErr, " + \
        "z_cModelMag, z_cModelMagErr, " + \
        "y_cModelMag, y_cModelMagErr " + \
        "FROM dp1.Object AS obj " + \
        "WHERE (i_cModelFlux/i_cModelFluxErr > 15) AND " + \
        "(u_cModelFlux/u_cModelFluxErr > 5) AND " + \
        "(shape_flag = 0) AND " + \
        "(g_extendedness = 0) AND " + \
        "(r_extendedness = 0) AND " + \
        "(i_extendedness = 0) AND " + \
        "(z_extendedness = 0) AND " + \
        "CONTAINS(POINT('ICRS', coord_ra, coord_dec), " + \
        "CIRCLE('ICRS',"+str(target_ra)+","+str(target_dec)+", "+str(target_r)+")) = 1 "



In [64]:
job = service.submit_job(query)
job.run()
job.wait(phases=['COMPLETED', 'ERROR'])
print('Job phase is', job.phase)

Job phase is COMPLETED


In [65]:
results = job.fetch_result()
tab = results.to_table()
tab

<Table length=2231>
     objectId      u_cModelMag u_cModelMagErr ... y_cModelMag y_cModelMagErr
                       mag          mag       ...     mag          mag      
      int64          float32      float32     ...   float32      float32    
------------------ ----------- -------------- ... ----------- --------------
611257134226610698     22.4926       0.101755 ...          --             --
611256515751336084      22.457       0.162976 ...          --             --
609788117972428529     17.8141     0.00402551 ...          --             --
609788117972428638     19.3623     0.00975272 ...          --             --
609788117972429619     21.6582      0.0553251 ...     16.2079     0.00131716
609789354923012483     20.7415      0.0106716 ...     15.2817     0.00138138
609788667728239946      23.397       0.101236 ...      18.752       0.015906
609788667728239945     17.3655     0.00139021 ...     15.6729     0.00173974
609788667728239912     20.8531      0.0150805 ...     15.6909      0.0017643
               ...         ...            ... ...         ...            ...
611256447031851094     22.2196      0.0414677 ...     17.3404     0.00167309
611256447031854552     19.3879     0.00638728 ...     15.7121     0.00117648
611256447031852188      21.158       0.016575 ...     20.0213      0.0262117
611256447031855028     23.0097       0.131655 ...     17.7743      0.0049478
611256447031855056     21.4899      0.0232077 ...     20.1325       0.027758
611256447031852314     23.2778       0.103845 ...     18.3669     0.00606183
611256447031851194     20.3382     0.00802921 ...     15.1378    0.000488173
611256447031854266      23.286       0.106414 ...     19.7633      0.0242488
611256447031852378     22.4862      0.0480762 ...     17.2222     0.00196879

In [66]:
job.delete()
del query, results, tab

## DC2

In [40]:
query = "SELECT mt.id_truth_type AS mt_id_truth_type, "\
        "mt.match_objectId AS mt_match_objectId, "\
        "ts.ra AS ts_ra, "\
        "ts.dec AS ts_dec, "\
        "ts.truth_type AS ts_truth_type, "\
        "ts.is_pointsource AS ts_is_pointsource, "\
        "ts.redshift AS ts_redshift, "\
        "obj.coord_ra AS obj_coord_ra, "\
        "obj.coord_dec AS obj_coord_dec, "\
        "obj.refExtendedness AS obj_refExtendedness, "\
        "scisql_nanojanskyToAbMag(obj.u_cModelFlux) AS umag, "\
        "scisql_nanojanskyToAbMag(obj.g_cModelFlux) AS gmag, "\
        "scisql_nanojanskyToAbMag(obj.r_cModelFlux) AS rmag, "\
        "scisql_nanojanskyToAbMag(obj.i_cModelFlux) AS imag, "\
        "scisql_nanojanskyToAbMag(obj.z_cModelFlux) AS zmag, "\
        "scisql_nanojanskyToAbMag(obj.y_cModelFlux) AS ymag, "\
        "scisql_nanojanskyToAbMagSigma(obj.g_cModelFlux, obj.u_cModelFluxErr) AS umag_err, "\
        "scisql_nanojanskyToAbMagSigma(obj.g_cModelFlux, obj.g_cModelFluxErr) AS gmag_err, "\
        "scisql_nanojanskyToAbMagSigma(obj.g_cModelFlux, obj.r_cModelFluxErr) AS rmag_err, "\
        "scisql_nanojanskyToAbMagSigma(obj.g_cModelFlux, obj.i_cModelFluxErr) AS imag_err, "\
        "scisql_nanojanskyToAbMagSigma(obj.g_cModelFlux, obj.z_cModelFluxErr) AS zmag_err, "\
        "scisql_nanojanskyToAbMagSigma(obj.g_cModelFlux, obj.y_cModelFluxErr) AS ymag_err "\
        "FROM dp02_dc2_catalogs.MatchesTruth AS mt "\
        "JOIN dp02_dc2_catalogs.TruthSummary AS ts ON mt.id_truth_type = ts.id_truth_type "\
        "JOIN dp02_dc2_catalogs.Object AS obj ON mt.match_objectId = obj.objectId "\
        "WHERE CONTAINS(POINT('ICRS', obj.coord_ra, obj.coord_dec), CIRCLE('ICRS', 62.0, -37.0, 0.5)) = 1 "\
        "AND scisql_nanojanskyToAbMag(obj.y_cModelFlux)/scisql_nanojanskyToAbMagSigma(obj.y_cModelFlux, obj.y_cModelFluxErr) > 5 "\
        "AND obj.detect_isPrimary = 1"

In [41]:
job = service.submit_job(query)
job.run()
job.wait(phases=['COMPLETED', 'ERROR'])
print('Job phase is', job.phase)

Job phase is COMPLETED


In [42]:
results = job.fetch_result().to_table().to_pandas()
print(len(results))

389077


In [43]:
results

mt_id_truth_type    mt_match_objectId      ts_ra     ts_dec  \
0          7940517773_1  1651413688361430386  61.929895 -36.523061   
1          7941123620_1  1651413688361451853  61.900438 -36.519216   
2          7941880022_1  1651413688361451734  61.869686 -36.521054   
3         31411981094_2  1651413688361421449  61.944366 -36.507034   
4         31500767183_2  1651413688361420776  61.926042 -36.520935   
...                 ...                  ...        ...        ...   
389072     7945365877_1  1651290543059148909  61.901237 -36.882347   
389073     7937556894_1  1651343319617242060  62.214016 -36.764291   
389074     7938263957_1  1651343319617244715  62.077497 -36.705900   
389075     7939442820_1  1651352115710265560  61.992653 -36.734153   
389076     7939683565_1  1651281746966093441  62.214007 -36.836013   

        ts_truth_type  ts_is_pointsource  ts_redshift  obj_coord_ra  \
0                   1                  0     1.079690     61.929825   
1                   1                  0     1.173830     61.900534   
2                   1                  0     1.301130     61.869544   
3                   2                  1     0.000000     61.944356   
4                   2                  1     0.000000     61.925989   
...               ...                ...          ...           ...   
389072              1                  0     1.819700     61.901248   
389073              1                  0     0.223015     62.214047   
389074              1                  0     0.624259     62.077490   
389075              1                  0     0.942635     61.992615   
389076              1                  0     1.000200     62.214027   

        obj_coord_dec  obj_refExtendedness  ...       rmag       imag  \
0          -36.523022                  1.0  ...  26.155623  25.903313   
1          -36.519245                  1.0  ...  26.445364  27.257660   
2          -36.521018                  1.0  ...  25.692458  25.894309   
3          -36.507118                  1.0  ...  25.961597  25.452401   
4          -36.520908                  1.0  ...  26.642154  25.257680   
...               ...                  ...  ...        ...        ...   
389072     -36.882354                  1.0  ...  25.208321  24.954217   
389073     -36.764303                  0.0  ...  25.648740  25.602706   
389074     -36.705880                  1.0  ...  25.562560  24.937279   
389075     -36.734149                  0.0  ...  26.414832  25.482749   
389076     -36.836029                  1.0  ...  25.052325  23.950500   

             zmag       ymag  umag_err  gmag_err  rmag_err  imag_err  \
0             NaN  25.071435  1.011092  0.344147  0.421003  0.789069   
1             NaN  25.378260  0.864431  0.278411  0.318666  0.617453   
2       24.946288        NaN  0.276842  0.089943  0.106828  0.209841   
3       25.984570  24.876525  0.742865  0.249135  0.296967  0.585449   
4       25.661938  24.325669  1.139720  0.343763  0.399996  0.766487   
...           ...        ...       ...       ...       ...       ...   
389072  24.842578  24.267990  0.120955  0.036944  0.041346  0.080011   
389073  25.165273  25.539307  0.324876  0.109264  0.125137  0.245820   
389074  24.869845  24.877324  0.359569  0.118269  0.142998  0.271518   
389075  25.095093  24.341662  0.684010  0.213506  0.255672  0.494887   
389076  23.418686  22.834803  0.569277  0.205507  0.269370  0.521795   

        zmag_err  ymag_err  
0       2.188033  3.702267  
1       1.674067  3.248848  
2       0.557984  1.067943  
3       1.546722  2.725028  
4       2.153979  3.881310  
...          ...       ...  
389072  0.226193  0.434724  
389073  0.677128  1.428278  
389074  0.677786  1.380787  
389075  1.322551  2.882764  
389076  1.211474  2.647471  

[389077 rows x 22 columns]